mino/router/mod.go

```diff
@@ -0,0 +1,53 @@
+// Package router defines the primitives to route a packet among a set of
+// participants.
+package router
+
+import (
+	"go.dedis.ch/dela/mino"
+	"go.dedis.ch/dela/serde"
+)
+
+// Packet is the type of message processed by the router. It contains
+// information that will allow the message to be routed.
+type Packet interface {
+	serde.Message
+
+	// GetSource returns the source address of the message.
+	GetSource() mino.Address
+
+	// GetDestination returns the destination address of the message.
+	GetDestination() []mino.Address
+
+	// GetMessage returns the message to be transmitted to the application when
+	// the source address is the current node. Message is only deserialized when
+	// needed.
+	GetMessage(ctx serde.Context, f serde.Factory) (serde.Message, error)
+}
+
+// Router is the interface of the routing service. It provides the primitives to
+// route a packet among a set of participants.
+type Router interface {
+	// MakePacket should be first called by the caller to set the specific
+	// required attribute on the packet if needed, for example a seed.
+	MakePacket(me mino.Address, to []mino.Address, msg []byte) Packet
+
+	// Forward takes the destination address, unmarshal the packet, and, based
+	// on its content, return a map of packets, where each element of the map
+	// represents the destination as the key, and the packet to send as the
+	// value. The simplest forward would be, if the destination is A,B,C, to
+	// create a map with 3 entries:
+	//
+	//  {A: packet{to: [A]}, B: packet{to: [B]}, C: packet{to: [C]}}.
+	//
+	// For a tree routing, the first message will send a packet with all the
+	// recipients to the root address (A in the following), like:
+	//
+	//	{A: packet{to: [A, B, C]}}
+	//
+	Forward(me mino.Address, data []byte, ctx serde.Context) (map[mino.Address]Packet, error)
```

Comment:
it should be `Forward(mino.Address, Packet) (...)`. Again the router does not care about encoding.

Reply:
The router can provide the packet factory.